### ISSUE TITLE
Question: How does this work with the newly release Ubuntu Store version?

### ISSUE THREAD

**Issue by klauern** [mintty/wsltty#52]
While I'm sure it's all of maybe a day old, I thought it'd be worthwhile to ask how this interoperates or will interoperate with the direction that seems to be going with WSL:

https://blogs.msdn.microsoft.com/commandline/2017/07/10/ubuntu-now-available-from-the-windows-store/

It appears that there's an official Ubuntu install in the Store, with other flavors forthcoming.  Additionally, it appears that the plan is to have them run side-by-side or simultaneously.

Most importantly, however, I saw this on the blog post's FAQ:

> Q2) Should I continue to run the current/legacy distro or move to the new Store distros
> A2) You can continue to run the current/legacy distro, but we do encourage you to move to using the store distro’s as your primary distro’s, because the current/legacy distro will be deprecated at some point.

With the "old" way of starting and installing the WSL subsystem with Ubuntu, how will this tool work?  How can I migrate it to use the `ubuntu` command that appears different/distinct from WSL's `bash` callout?

**Comment by nevhan** [comment]
```
C:\Users\nevhan>where ubuntu
C:\Users\nevhan\AppData\Local\Microsoft\WindowsApps\ubuntu.exe

C:\Users\nevhan>where bash
C:\Windows\System32\bash.exe

C:\Users\nevhan>ver
Microsoft Windows [Version 10.0.16232.1000]
```

I have uninstalled the legacy environment and it appears bash opens the new windows store version of ubuntu. WSLTTY still works but it complains about the icon being missing.

**Comment by mintty** [comment]
I don't have the update yet, but if you tell me the new path name of the icon file, I can change installation so that it copies either of them into the wsltty installation directory (rather than refer to the assumed location).
As a workaround, you can configure the new icon location in the config file (Icon=).

**Comment by mintty** [comment]
Checking the announcement linked above:
> * Install distros side-by-side: WSL has been enhanced to support the installation of multiple distro’s alongside one another
> * Run multiple distro’s simultaneously: Not only can you install distro’s side-by-side, but you can now run more than one distro simultaneously

If anyone can contribute an idea how to handle that with wsltty, and how to find out the actual installation location(s), I'd appreciate it.

**Comment by mintty** [comment]
> A3) You can launch the new store based distro’s via the start menu tile created for each distro.

Information might be extracted here. Also wsltty could semi-automatically add an invocation shortcut per distro, assuming they can be recognized and distinguished.

**Comment by klauern** [comment]
I took a look last night to see if I could find out the location of the Ubuntu install, but was unsuccessful.  I'll try again tonight.  The usual "Open File Location" wasn't available for this app, and even after pinning it to the Start menu, I wasn't seeing it.  Any tips on where to get the executable location from this would be appreciated.

**Comment by mintty** [comment]
Wouldn't @nevhan's tip above

> C:\Users\nevhan>where ubuntu
> C:\Users\nevhan\AppData\Local\Microsoft\WindowsApps\ubuntu.exe

be a good starting point to look?

**Comment by klauern** [comment]
```ps1
C:\Users\klauer
λ  which ubuntu

Definition
----------
C:\Users\klauer\AppData\Local\Microsoft\WindowsApps\ubuntu.exe

```

**Comment by klauern** [comment]
![image](https://user-images.githubusercontent.com/4735/28187338-ff6e25b0-67e3-11e7-9700-ec85386ecaff.png)

**Comment by klauern** [comment]
Similarly, inside the `Canonical...` directory, is another 0-byte `ubuntu.exe`.

**Comment by keyduq** [comment]
In this directory are store everything about it but cant find any icon file
```C:\Users\{user}\AppData\Local\Packages\CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc```
Rootfs directory: ```C:\Users\{user}\AppData\Local\Packages\CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc\LocalState\rootfs```

**Comment by mintty** [comment]
What about the Start Menu icon (after "pinning"), can you open "Properties"? It should expose target, icon, etc. That information would be needed anyway in order to find out how to address the different parallel WSL installations that are now possible.

About the icon, as the MS pages annonce that other Linux distributions than Ubuntu will be available shorty, maybe I should switch to a more generic icon anyway (Tux).

**Comment by klauern** [comment]
Here's what I get from the start menu icon after pinning:

![image](https://user-images.githubusercontent.com/4735/28229865-265adff8-68aa-11e7-883e-235af64e4474.png)

**Comment by mintty** [comment]
What about the "normal" Start Menu icon (not the tile)?

**Comment by klauern** [comment]
Same:

![image](https://user-images.githubusercontent.com/4735/28230213-a328c4d6-68ab-11e7-92cd-4617bf84a017.png)

As is the "Pin to Task Bar" options:

![image](https://user-images.githubusercontent.com/4735/28230244-b8eb0450-68ab-11e7-8cc4-5c11a2df5315.png)

**Comment by klauern** [comment]
I wonder if this is hidden in the registry somewhere because this seems a little too-well hidden to be this hard to find.

**Comment by mintty** [comment]
You should be able to find some usable Start Menu shortcut files under `%APPDATA%\Roaming\Microsoft\Windows\Start Menu` or `%ProgramData%\Microsoft\Windows\Start Menu`.

**Comment by klauern** [comment]
I might have found something.  I think this path is different depending on who installs it, because it looks auto-generated to me, but the path under  `C:\Program Files\WindowsApps\CanonicalGroupLimited.UbuntuonWindows_1604.2017.711.0_x64__79rhkp1fndgsc\Assets` has a bunch of icons and such:

![image](https://user-images.githubusercontent.com/4735/28236145-b6a40024-68e3-11e7-8697-82169ddb4404.png)

**Comment by klauern** [comment]
Unfortunately, I can't run the `ubuntu.exe` natively from that path, as it gives me some kind of permission denied error:

![image](https://user-images.githubusercontent.com/4735/28236154-f3d44378-68e3-11e7-85a0-0a6d241bd610.png)


![image](https://user-images.githubusercontent.com/4735/28236152-ea5d9362-68e3-11e7-8298-fa7447c229da.png)

The way I got this was by right-clicking on the Ubuntu app when it starts up and seeing the properties for it:

![image](https://user-images.githubusercontent.com/4735/28236158-1f0bb1ca-68e4-11e7-8c5b-9ae7601b6063.png)

![image](https://user-images.githubusercontent.com/4735/28236161-279a0558-68e4-11e7-9f7c-679c7d6eb200.png)


I'm not sure how it's getting called, though, because it doesn't seem to work without running from the Start menu.

**Comment by klauern** [comment]
Simpler yet, under the `images/` subfolder is one file called `icon.ico` that looks like it might be it:

![image](https://user-images.githubusercontent.com/4735/28236173-735de1a8-68e4-11e7-9d1d-3566a64221dc.png)

**Comment by mintty** [comment]
I have a script now that would extract distro information and icons if possible (but only if run as admin and icon is at a standard location, like with Ubuntu, as Microsoft does not provide information about how to get an icon from a hidden Start Menu entry).
The script is now a shell script and needs some transformation to avoid bundling a bunch of DLL dependencies.
However (@rprichard), we still need support in the wslbridge frontend to use a distinct launcher rather than bash.exe.

**Comment by rprichard** [comment]
I'll try to add support for a custom launcher path in a few days.

**Comment by mintty** [comment]
@rprichard, thanks; I also tried to patch wslbridge but my first attempt failed. See also https://github.com/rprichard/wslbridge/issues/14#issuecomment-318000272 please.

**Comment by mintty** [comment]
Released 1.7.9. Please check, feedback appreciated.

**Comment by klauern** [comment]
Is there an `installer.exe` that we can use like the other versions?  If not, what's the process to building this for testing?

**Comment by mintty** [comment]
I had forgotten to attach the installer, sorry. Now available.
Otherwise, you could build it simply by downloading the source archive and running `make`, within a cygwin installation.
Also, then you could run `make install` to install directly, without the installer.

**Comment by klauern** [comment]
Excellent! I was able to test it and I think it works.  I have both the old WSL install and Ubuntu, so it has context items for both.  I was waiting on this to work before uninstalling the old WSL distro, so now that I have it, I'm pretty stoked.

If anyone else wants to chime in on their tests, feel free, but for me, this looks like it works so I would be comfortable closing the issue.

@mintty @rprichard thank you for working on this!

**Comment by mintty** [comment]
Thank you.
Actually I still see an issue; if you uninstall your old WSL, its Explorer context menu entry will persist, and the "remove..." script will not remove it anymore. Maybe I should add a marker in the registry for that purpose.

**Comment by mintty** [comment]
Released 1.8.0 with full support for a variety of WSL distributions, and the WSL default distribution.

**Comment by mintty** [comment]
See #63 about my request for a Windows Store package.

**Comment by kbui710** [comment]
Hi all, I got the error below when try to run Ubuntu on Windows 10. I have checked the path C:\Program Files\WindowsApps\... to see if its valid but there is no such folder 'WindowsApps' under 'Program Files'. Anyone help is much appreciated. 
<img width="414" alt="ubt1" src="https://user-images.githubusercontent.com/36374871/36075377-bce96d04-0f80-11e8-98b6-46c91b79a1a2.png">
<img width="396" alt="ubt2" src="https://user-images.githubusercontent.com/36374871/36075417-5d07dde8-0f81-11e8-8e00-3cdff5b9758b.png">

**Comment by willpower232** [comment]
@kbui710 please open a new issue as your problem is probably something different.

FWIW, WindowsApps is a hidden folder so you'll need to reveal those in explorer to see it. Please also make sure you have the latest version of wsltty installed.

**Comment by kbui710** [comment]
@willpower232 thank you for the reply and helpful information.

**Comment by mintty** [comment]
Actually, wsltty does not run the ubuntu.exe wrapper so I doubt you were really using wsltty at all.
If you open another issue, please describe precisely what you were doing.
